ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,200 @@
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    event('Query Cache successful Get', {context: 'queryCache', queryID: id})
+    return this.cache[id].values
+  }
+
+  initializeCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    isCustomTime: boolean = false
+  ) => {
+    if (this.cache[queryID]) {
+      return this.cache[queryID]
+    }
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      hashedVariables,
+      isCustomTime,
+      mutex: RunQueryPromiseMutex<RunQueryResult>(),
+    }
+    return this.cache[queryID]
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    values: RunQueryResult
+  ): void => {
+    event('Query Cache was Set', {context: 'queryCache', queryID})
+    this.cache[queryID] = {
+      ...this.initializeCacheByID(queryID, hashedVariables),
+      dateSet: Date.now(),
+      values,
+    }
+  }
+
+  startWatchDog = () => {
+    setInterval(() => {
+      this.cleanExpiredQueries()
+    }, TIME_INVALIDATION / 2)
+
+    this.cleanExpiredQueries()
+  }
+}
+
+const queryCache = new QueryCache()
+// Set an interval to check for expired data to invalidate
+queryCache.startWatchDog()
+
+export const resetQueryCache = (): void => {
+  queryCache.resetCache()
+}
+
+export const resetQueryCacheByQuery = (query: string): void => {
+  const queryID = `${hashCode(query)}`
+  queryCache.resetCacheByID(queryID)
+}
+
+export const getCachedResultsOrRunQuery = (
+  orgID: string,
+  query: string,
+  state: AppState
+): CancelBox<RunQueryResult> => {
+  const queryID = `${hashCode(query)}`
+  event('Starting Query Cache Process ', {context: 'queryCache', queryID})
+  const allVars = getAllVariables(state)
+  const usedVars = filterUnusedVarsBasedOnQuery(allVars, [query])
+  const variables = sortBy(usedVars, ['name'])
+  const simplifiedVariables = variables.map(v => asSimplyKeyValueVariables(v))
+  const stringifiedVars = JSON.stringify(simplifiedVariables)
+  // create the queryID based on the query & vars
+  const hashedVariables = `${hashCode(stringifiedVars)}`
+
+  const cacheResults: RunQueryResult | null = queryCache.getFromCache(
+    queryID,
+    hashedVariables
+  )
+
+  // check the cache based on text & vars
+  if (cacheResults) {
+    return {
+      promise: new Promise(resolve => resolve(cacheResults)),
+      cancel: () => {},
+    }
+  }
+  const variableAssignments = variables
+    .map(v => asAssignment(v))
+    .filter(v => !!v)
+
+  const windowVars = getWindowVars(query, allVars)
+
+  // otherwise query & set results
+  const extern = buildVarsOption([...variableAssignments, ...windowVars])
```

Comment:
flagging this for the thread's sake as the fix for the outage... generating a `windowPeriod` variable from the query and adding it to the AST partial for the query

Reply:
👍